lib/configurators/aws/aws.go

```diff
@@ -883,6 +886,31 @@ func buildSecretsManagerStatements(fileConfig *config.FileConfig, target awslib.
 	return statements
 }
 
+// buildSTSAssumeRoleStatements returns AWS IAM statements necessary for
+// assuming AWS IAM roles.
+func buildSTSAssumeRoleStatements(flags configurators.BootstrapFlags, targetCfg targetConfig, boundary bool) ([]*awslib.Statement, error) {
+	if boundary {
+		// for boundary policy, we need to add sts permission if any assume role is configured.
+		if !flags.ForceBoundaryAssumeRoles && targetCfg.assumeAWSRoles.len() == 0 {
+			return nil, nil
+		}
+		return []*awslib.Statement{{
+			Effect:    awslib.EffectAllow,
+			Actions:   stsActions,
+			Resources: []string{"*"},
+		}}, nil
+	}
+	// for permissions policy, we only need to add permissions for external roles.
+	if len(targetCfg.assumeAWSRoles.external) == 0 {
+		return nil, nil
+	}
```

Comment:
internal roles as in "internal within the target identity's account". Assuming an external role has stricter permissions requirements - the role must trust the identity and the identity must have permissions to assume the role. For this reason, we need to include external roles in the permissions policy like:
```
        {
            "Effect": "Allow",
            "Action": "sts:AssumeRole",
            "Resource": "arn:aws:iam::123456789012:role/role-1"
        },
```
For roles within the same account ("internal" roles), it's enough to just have the trust policy and allow sts:AssumeRole for `*` in the permission boundary for the target identity.

Reply:
I'll add a link to AWS docs in the godoc for reference

Reply:
> For roles within the same account ("internal" roles), it's enough to just have the trust policy and allow sts:AssumeRole for * in the permission boundary for the target identity.

Correct me if I am wrong, `sts:AssumeRole` is not needed if the trust policy explicitly allows the assumer role. If trust policy is the [current account ID](https://goteleport.com/docs/database-access/guides/aws-cassandra-keyspaces/#step-35-create-an-aws-keyspaces-role), `sts:AssumeRole` is still required?

Reply:
> Correct me if I am wrong, sts:AssumeRole is not needed if the trust policy explicitly allows the assumer role.

That's right, but we still need sts:AssumeRole in the permissions boundary or it will be denied.


https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html#policies_resource-based
>> To enable cross-account access, you can specify an entire account or IAM entities in another account as the principal in a resource-based policy. Adding a cross-account principal to a resource-based policy is only half of establishing the trust relationship. When the principal and the resource are in separate AWS accounts, you must also use an identity-based policy to grant the principal access to the resource. However, if a resource-based policy grants access to a principal in the same account, no additional identity-based policy is required.


> And i don't think external_id is a good indicator it's external role.

I'll check the target identity's account ID compared to assume_role_arn instead.

> For simplicity, maybe we can just always include the sts action to assume these roles.

I didn't want to include "internal" roles in the sts permissions statements because policy size is limited, and there's no need for those roles to be allowed explicitly. I also didn't want to allow `*` in the permissions policy (even though we do in the boundary), so I settled on only including external roles in the permissions policy and `*` in the boundary.

Reply:
I think what I meant is depending on how the trust policy is set up, `sts:AssumeRole` is still required in the policy for the role in the same account (regardless of the boundry).

I have setup two roles in our teleport-dev account: `steve-assume-test` and `steve-trust-account`. The trust relationship is setup using the "AWS Account" option when creating the role (https://goteleport.com/docs/database-access/guides/aws-cassandra-keyspaces/#step-35-create-an-aws-keyspaces-role), and it looks like this:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": "arn:aws:iam::<account-number>:root"
            },
            "Action": "sts:AssumeRole",
            "Condition": {}
        }
    ]
}
```
In this case, even `steve-assume-test` is in the same account, `steve-assume-test` still needs explicit `sts:AssumeRole` policy otherwise it cannot assume `steve-trust-account`.

I don't mind excluding the assume role action for "internal" roles as long as we document how to set up trust relationships very carefully in our docs. Just I think always including the action is simpler and more effective.

Reply:
oh, I see what you meant now. Yes it would be convenient to add those "internal" roles just in case their trust policy doesn't specify the target identity explicitly. I'll eliminate the distinction for internal/external then. I suppose I should go ahead and change `--assumes-external-roles` into `--assumes-roles` as well, since the distinction between internal/external won't matter anymore.

It's still not perfect though: For keyspaces/dynamo/redshiftserverless, the configurator doesn't know which roles need to be put in the permissions policy. For aws cross-account, it *can* know the ones in static config, but it won't be aware of roles from dynamic databases created by `tctl` or via discovery service. I guess they can use the cli flag to include those roles though. I'll make sure to mention this in the docs PR for aws cross account, and update other guides as needed too.

Reply:
another option is the assume role action can use conditions on resource tags. 
```
        {
            "Effect": "Allow",
            "Action": [
                "sts:AssumeRole"
            ],
            "Resource": "*",
            "Condition": {
                "StringEquals": {
                    "aws:ResourceTag/TeleportAssumeRole": "true"
                }
            }
        }
```

Reply:
I'd rather not add the additional setup of tagging roles. I reworked it to just add all the roles to the sts permissions